File: main-workshop/services/dotnet/ride-service/Services/RideService.cs
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;
using Amazon.EventBridge;
using Amazon.EventBridge.Model;
using System.Text.Json;

[assembly: LambdaSerializer(typeof(Amazon.Lambda.Serialization.SystemTextJson.DefaultLambdaJsonSerializer))]

public class RideService
{
    private readonly IAmazonDynamoDB _dynamoDb;
    private readonly IAmazonEventBridge _eventBridge;
    private readonly string _tableName;
    private readonly string _eventBusName;
    private readonly JsonSerializerOptions _serializeOptions;

    public RideService()
    {
        _dynamoDb = new AmazonDynamoDBClient();
        _eventBridge = new AmazonEventBridgeClient();
        _tableName = Environment.GetEnvironmentVariable("RIDES_TABLE_NAME") ?? "Rides";
        _eventBusName = Environment.GetEnvironmentVariable("EVENT_BUS_NAME") ?? "";
        
        _serializeOptions = new JsonSerializerOptions
        {
            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
            WriteIndented = false
        };
    }

    public async Task<RideCreationResult> CreateRideAsync(APIGatewayProxyRequest request, string? deviceId = null)
    {
        var result = new RideCreationResult();

        try
        {
            // Extract correlation ID from request headers
            var correlationId = request.Headers?.GetValue("x-correlation-id");
            if (string.IsNullOrEmpty(correlationId))
            {
                correlationId = null;
            }

            // Deserialize request
            var createRequest = JsonSerializer.Deserialize<CreateRideRequest>(request.Body, _serializeOptions);
            if (createRequest == null)
            {
                result.Success = false;
                result.ErrorType = "InvalidRequest";
                result.ErrorMessage = "Invalid request format";
                return result;
            }

            // Create ride object
            var ride = new Ride
            {
                RideId = Guid.NewGuid().ToString(),
                RiderId = createRequest.RiderId,
                RiderName = createRequest.RiderName,
                PickupLocation = createRequest.PickupLocation,
                DestinationLocation = createRequest.DestinationLocation,
                PaymentMethod = createRequest.PaymentMethod,
                DeviceId = deviceId, // Store device ID with the ride
                Status = "requested",
                CreatedAt = DateTime.UtcNow.ToString("O"),
                UpdatedAt = DateTime.UtcNow.ToString("O")
            };

            // Save to DynamoDB
            await SaveRideToDynamoDbAsync(ride);

            // Send event to EventBridge with correlation ID
            await SendRideCreatedEventAsync(ride, correlationId);

            result.Success = true;
            result.Ride = ride;
            return result;
        }
        catch (JsonException)
        {
            result.Success = false;
            result.ErrorType = "JsonException";
            result.ErrorMessage = "Invalid JSON format";
            return result;
        }
        catch (Exception ex)
        {
            result.Success = false;
            result.ErrorType = "UnexpectedError";
            result.ErrorMessage = ex.Message;
            throw;
        }
    }

    private async Task SaveRideToDynamoDbAsync(Ride ride)
    {
        var item = new Dictionary<string, AttributeValue>
        {
            { "rideId", new AttributeValue { S = ride.RideId } },
            { "riderId", new AttributeValue { S = ride.RiderId } },
            { "riderName", new AttributeValue { S = ride.RiderName } },
            { "pickupLocation", new AttributeValue { S = JsonSerializer.Serialize(ride.PickupLocation) } },
            { "destinationLocation", new AttributeValue { S = JsonSerializer.Serialize(ride.DestinationLocation) } },
            { "paymentMethod", new AttributeValue { S = ride.PaymentMethod } },
            { "deviceId", new AttributeValue { S = ride.DeviceId ?? "unknown" } },
            { "status", new AttributeValue { S = ride.Status } },
            { "createdAt", new AttributeValue { S = ride.CreatedAt } },
            { "updatedAt", new AttributeValue { S = ride.UpdatedAt } }
        };

        var putRequest = new PutItemRequest
        {
            TableName = _tableName,
            Item = item
        };

        await _dynamoDb.PutItemAsync(putRequest);
    }

    private async Task SendRideCreatedEventAsync(Ride ride, string correlationId)
    {
        if (string.IsNullOrEmpty(_eventBusName))
        {
            return;
        }

        var rideCreatedEvent = new RideCreatedEvent
        {
            RideId = ride.RideId,
            RiderId = ride.RiderId,
            RiderName = ride.RiderName,
            PickupLocation = ride.PickupLocation,
            DestinationLocation = ride.DestinationLocation,
            PaymentMethod = ride.PaymentMethod,
            Timestamp = DateTime.UtcNow,
            EventType = "RideCreated",
            CorrelationId = correlationId
        };

        var eventDetailJson = JsonSerializer.Serialize(rideCreatedEvent, _serializeOptions);

        var putEventsRequest = new PutEventsRequest
        {
            Entries = new List<PutEventsRequestEntry>
            {
                new PutEventsRequestEntry
                {
                    Source = "ride-service",
                    DetailType = "RideCreated",
                    Detail = eventDetailJson,
                    EventBusName = _eventBusName
                }
            }
        };

        var result = await _eventBridge.PutEventsAsync(putEventsRequest);
            
        var failedEntries = result.Entries.Where(e => !string.IsNullOrEmpty(e.ErrorCode)).ToList();
        if (failedEntries.Any())
        {
            throw new Exception($"Failed to send event: {failedEntries.First().ErrorCode}");
        }
    }
}

public static class Extensions
{
    public static string GetValue(this IDictionary<string,string> dict, string key)
    {
        return dict.TryGetValue(key, out var value) ? value : string.Empty;
    }
}

File: main-workshop/services/typescript/ride-service/src/services/RideService.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import {
  EventBridgeClient,
  PutEventsCommand,
} from '@aws-sdk/client-eventbridge';
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';
import type { APIGatewayProxyEvent } from 'aws-lambda';
import { captureAWSv3Client } from 'aws-xray-sdk-core';
import { randomUUID } from 'crypto';
import type { CreateRideRequest, Ride, RideCreatedEvent } from '../models';

export interface RideCreationResult {
  success: boolean;
  ride?: Ride;
  errorType?: string;
  errorMessage?: string;
}

export class RideService {
  private readonly docClient: DynamoDBDocumentClient;
  private readonly tableName: string;
  private readonly eventBridge: EventBridgeClient;
  private readonly eventBusName: string;

  constructor() {
    const client = new DynamoDBClient({});
    this.docClient = captureAWSv3Client(DynamoDBDocumentClient.from(client));
    this.tableName = process.env.RIDES_TABLE_NAME || 'Rides';
    this.eventBridge = captureAWSv3Client(new EventBridgeClient({}));
    this.eventBusName = process.env.EVENT_BUS_NAME || '';
  }

  async createRide(
    event: APIGatewayProxyEvent,
    deviceId?: string
  ): Promise<RideCreationResult> {
    try {
      // Extract correlation ID from request headers
      const correlationId = this.getHeaderValue(
        event.headers,
        'x-correlation-id'
      );

      // Validate request body
      const requestBody = event.body;
      if (!requestBody) {
        return {
          success: false,
          errorType: 'InvalidRequest',
          errorMessage: 'Request body is required',
        };
      }

      // Parse request body
      let request: CreateRideRequest;
      try {
        request = JSON.parse(requestBody) as CreateRideRequest;
      } catch (error) {
        return {
          success: false,
          errorType: 'JsonException',
          errorMessage: 'Invalid JSON format',
        };
      }

      // Create ride object
      const ride: Ride = {
        rideId: randomUUID(),
        riderId: request.riderId,
        riderName: request.riderName,
        pickupLocation: request.pickupLocation,
        destinationLocation: request.destinationLocation,
        paymentMethod: request.paymentMethod,
        deviceId: deviceId,
        status: 'requested',
        createdAt: new Date().toISOString(),
        updatedAt: new Date().toISOString(),
      };

      // Save to DynamoDB
      await this.saveRideToDynamoDB(ride);

      // Send event to EventBridge
      await this.sendRideCreatedEvent(ride, correlationId);

      return {
        success: true,
        ride,
      };
    } catch (error) {
      return {
        success: false,
        errorType: 'UnexpectedError',
        errorMessage: error instanceof Error ? error.message : 'Unknown error',
      };
    }
  }

  private async saveRideToDynamoDB(ride: Ride): Promise<void> {
    const item = {
      rideId: ride.rideId,
      riderId: ride.riderId,
      riderName: ride.riderName,
      pickupLocation: JSON.stringify(ride.pickupLocation),
      destinationLocation: JSON.stringify(ride.destinationLocation),
      paymentMethod: ride.paymentMethod,
      deviceId: ride.deviceId || 'unknown',
      status: ride.status,
      createdAt: ride.createdAt,
      updatedAt: ride.updatedAt,
    };

    const command = new PutCommand({
      TableName: this.tableName,
      Item: item,
    });

    await this.docClient.send(command);
  }

  private async sendRideCreatedEvent(
    ride: Ride,
    correlationId?: string
  ): Promise<void> {
    if (!this.eventBusName) {
      return;
    }

    const rideCreatedEvent: RideCreatedEvent = {
      rideId: ride.rideId,
      riderId: ride.riderId,
      riderName: ride.riderName,
      pickupLocation: ride.pickupLocation,
      destinationLocation: ride.destinationLocation,
      paymentMethod: ride.paymentMethod,
      timestamp: new Date().toISOString(),
      eventType: 'RideCreated',
      correlationId,
    };

    const command = new PutEventsCommand({
      Entries: [
        {
          Source: 'ride-service',
          DetailType: 'RideCreated',
          Detail: JSON.stringify(rideCreatedEvent),
          EventBusName: this.eventBusName,
        },
      ],
    });

    const result = await this.eventBridge.send(command);

    const failedEntries = result.Entries?.filter((e: any) => e.ErrorCode);
    if (failedEntries && failedEntries.length > 0) {
      throw new Error(`Failed to send event: ${failedEntries[0].ErrorCode}`);
    }
  }

  private getHeaderValue(
    headers: { [key: string]: string | undefined } | null,
    headerName: string
  ): string | undefined {
    if (!headers) {
      return undefined;
    }

    const headerEntry = Object.entries(headers).find(
      ([key]) => key.toLowerCase() === headerName.toLowerCase()
    );

    return headerEntry?.[1];
  }
}
